modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java

```diff
@@ -206,40 +214,59 @@ public CompletableFuture<Void> notifyWatches(long newRevision, List<Entry> updat
+
+            if (NOTIFICATION_FUTURE_UPDATER.compareAndSet(this, oldNotificationFuture, newNotificationFuture)) {
+                oldNotificationFuture.whenComplete(copyStateTo(chainingFuture));
 
-        CompletableFuture<Void> newFuture = notificationFuture
-                .thenComposeAsync(v -> inBusyLockAsync(() -> {
-                    List<Entry> filteredUpdatedEntries = updatedEntries.isEmpty() ? emptyList() : updatedEntries.stream()
-                            .filter(WatchProcessor::isNotIdempotentCacheCommand)
-                            .collect(toList());
+                return newNotificationFuture;
+            }
+        }
+    }
 
-                    List<WatchAndEvents> watchAndEvents = collectWatchesAndEvents(filteredUpdatedEntries, newRevision);
+    private CompletableFuture<Void> notifyWatchesInternal(long newRevision, List<Entry> updatedEntries, HybridTimestamp time) {
+        assert time != null;
 
-                    long startTimeNanos = System.nanoTime();
+        return enqueue(() -> {
+            List<Entry> filteredUpdatedEntries = updatedEntries.isEmpty() ? emptyList() : updatedEntries.stream()
+                    .filter(WatchProcessor::isNotIdempotentCacheCommand)
+                    .collect(toList());
 
-                    CompletableFuture<Void> notifyWatchesFuture = performWatchesNotifications(watchAndEvents, newRevision, time);
+            List<WatchAndEvents> watchAndEvents = collectWatchesAndEvents(filteredUpdatedEntries, newRevision);
 
-                    // Revision update is triggered strictly after all watch listeners have been notified.
-                    CompletableFuture<Void> notifyUpdateRevisionFuture = notifyUpdateRevisionListeners(newRevision);
+            long startTimeNanos = System.nanoTime();
 
-                    CompletableFuture<Void> notificationFuture = allOf(notifyWatchesFuture, notifyUpdateRevisionFuture)
-                            .thenRunAsync(() -> inBusyLock(() -> invokeOnRevisionCallback(newRevision, time)), watchExecutor);
+            CompletableFuture<Void> notifyWatchesFuture = performWatchesNotifications(watchAndEvents, newRevision, time);
 
-                    notificationFuture.whenComplete((unused, e) -> maybeLogLongProcessing(filteredUpdatedEntries, startTimeNanos));
+            // Revision update is triggered strictly after all watch listeners have been notified.
+            CompletableFuture<Void> notifyUpdateRevisionFuture = notifyUpdateRevisionListeners(newRevision);
 
-                    return notificationFuture;
-                }), watchExecutor)
-                .whenComplete((unused, e) -> {
-                    if (e != null) {
-                        notifyFailureHandlerOnFirstFailureInNotificationChain(e);
-                    }
-                });
+            CompletableFuture<Void> notificationFuture = allOf(notifyWatchesFuture, notifyUpdateRevisionFuture)
```

Comment:
Would it make sense to give this variable another name (like `newNotificationFuture`) to make it more difficult to confuse it with `this.notificationFuture`?

Reply:
Yes, why not